### PR TITLE
Change Configuration Documentation

### PR DESCRIPTION
**Pull request by eljobe**
### Context
We need to do a better job explaining how configurations work, and specifically to encourage users to migrate from the Java Plugin to the Java Library Plugin as appropriate.

See background at: #1854

### CODE REVIEW

**Comment by eljobe** [comment]
@oehme and @adammurdoch, I'm really not that crazy about the changes I have added in this Pull Request, but it's hard to figure out exactly the right amount of detail to explain at this point in our development of these newer configurations.

One thing I've considered, but not yet done is the idea of adding another graphic to show the differences between the old and new configuration hierarchies. This graphic would be specifically targeting users who are looking for an easy migration path.

Please advise me on how to proceed.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 1]
```diff
@@ -378,7 +378,10 @@
                 <td>compile</td>
                 <td>-</td>
                 <td>-</td>
-                <td>Compile time dependencies</td>
+                <td>
+                    Compile time dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code> configuration defined
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: ❌ We should add `implementation`, `testImplementation` and `runtimeOnly` to this list instead of linking to the java-library plugin. These configurations are available in the `java` plugin already. Instead, we can either link back here from `java-library` or just duplicate it in both places.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 2]
```diff
@@ -396,7 +399,10 @@
                 <td>runtime</td>
                 <td>compile</td>
                 <td>-</td>
-                <td>Runtime dependencies</td>
+                <td>Runtime dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code>, <code>runtimeElements</code>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: ❌ `runtimeOnly` is the replacement for `runtime`, not `implementation`. `implementation is visible at compile time and is one of the replacements for `compile` (the other being `api` in the `java-library` plugin.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 3]
```diff
@@ -420,7 +426,9 @@
                 <td>testRuntime</td>
                 <td>runtime, testCompile</td>
                 <td>test</td>
-                <td>Additional dependencies for running tests only</td>
+                <td>Additional dependencies for running tests only. This configuration is weakly deprecated in favor of the <code>testImplementation</code> or <code>testRuntimeClasspath</code>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: `testRuntimeOnly`, same as above above.

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml [hunk 4]
```diff
@@ -71,44 +71,68 @@
...
-                <term>testCompile</term>
+                <term>runtimeClasspath</term>
                 <listitem>
                     <para>
-                        The dependencies required to compile the test source of the project. By default, also includes
-                        the compiled production classes and the compile time dependencies.
+                        The dependencies required by the production classes at runtime. By default, this includes the
+                        dependencies declared in the <code>api</code>, <code>implementation</code>, and <code>runtimeOnly</code> configurations.
+                        This configuration is an example of a configuration used for Resolving Dependencies, and as such, it should be unusual for
+                        users to declare dependencies directly in the <code>runtimeClasspath</code> configuration.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: They should *never* add dependencies to `runtimeClasspath`.

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml [hunk 5]
```diff
@@ -71,44 +71,68 @@
 
     <section id="configurations">
         <title>Dependency configurations</title>
-        <para>In Gradle dependencies are grouped into <firstterm>configurations</firstterm>. A configuration is simply a named set of dependencies.
-            We will refer to them as <firstterm>dependency configurations</firstterm>. You can use them to declare the external dependencies of your project.
-            As we will see later, they are also used to declare the publications of your project.</para>
-        <para>
-            The Java plugin defines a number of standard configurations. These configurations represent the classpaths that the
-            Java plugin uses. Some are listed below, and you can find more details in <xref linkend="tab:configurations"/>.
+        <para>In Gradle dependencies are grouped into <firstterm>configurations</firstterm>. A configuration is a named set of dependencies.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: I'd shorten this to just 

"A Configuration is a named set of dependencies and artifacts. There are three main purposes for a Configuration:"

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 1 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 2 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 3 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml (hunk 5 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: Done

**Comment by eljobe** [comment]
I think I addressed your comments @oehme, but please take another look. I'm also not sure I updated the right image. But, I think I did.

One thing I find confusing about all of this is that some of the configurations are put in place by the Java Plugin while others (just `api` and `apiElements`?) are added by  the Java Library Plugin. It's hard, for example, to explain on the Java Plugin page that the `implementation` configuration extends the `api` configuration when we don't really mention the `api` configuration in the rest of the document.

**Comment by oehme** [comment]
> t's hard, for example, to explain on the Java Plugin page that the implementation configuration extends the api configuration when we don't really mention the api configuration in the rest of the document.

We shouldn't mention that `extends` relationship in the `java` plugin guide, because `api` doesn't exist in the `java` plugin. We should just tell users "if your project is a library and needs to expose api dependencies to clients, use the `java-library` plugin.

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml [hunk 6]
```diff
@@ -71,44 +71,61 @@
...
+                    <term>Resolving Dependencies</term>
+                    <listitem>
+                        <para>The plugin uses configurations to find (and possibly download) the files it needs from outside the project to perform the work of its tasks,
+                            and the tasks of plugins which coordinate with the plugin by using the configurations it defines to reason about the project.</para>
+                    </listitem>
+                </varlistentry>
+                <varlistentry>
+                    <term>Exposing Artifacts for Consumption</term>
+                    <listitem>
+                        <para>The plugin uses configurations to define what <firstterm>artifacts</firstterm> it generates for other projects or subprojects to consume.</para>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: There is no difference between "projects" and "subprojects

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml [hunk 7]
```diff
@@ -71,44 +71,61 @@
...
+                    <term>Declaring Dependencies</term>
+                    <listitem>
+                        <para>The plugin uses configurations to make it easy for build authors to declare what other subprojects or external artifacts are needed for various purposes
+                            during the execution of tasks defined by the plugin.</para>
+                    </listitem>
+                </varlistentry>
+                <varlistentry>
+                    <term>Resolving Dependencies</term>
+                    <listitem>
+                        <para>The plugin uses configurations to find (and possibly download) the files it needs from outside the project to perform the work of its tasks,
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: I find this sentence hard to follow, especially the "coordinate and reason" part. The essence should be "These are used as task inputs".

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml [hunk 8]
```diff
@@ -71,44 +71,61 @@
...
                 </listitem>
             </varlistentry>
             <varlistentry>
-                <term>testRuntime</term>
+                <term>apiElements</term>
                 <listitem>
                     <para>
-                        The dependencies required to run the tests. By default, also includes the compile, runtime and test compile
-                        dependencies.
+                        The external dependencies which are part of this project's externally consumable API as well as the classes which are defined in this project which should be
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: remove "external"

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 9]
```diff
@@ -375,10 +375,21 @@
...
+                </td>
+            </tr>
+            <tr>
                 <td>compile</td>
                 <td>-</td>
                 <td>-</td>
-                <td>Compile time dependencies</td>
+                <td>
+                    Compile time dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code> configuration defined
+                    in <xref linkend="sec:java_library_configurations_graph"/>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: We should remove the reference to the java library plugin, `implementation` is described just a few lines above.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 10]
```diff
@@ -440,7 +472,7 @@
         <figure>
             <title>Java plugin - dependency configurations</title>
             <imageobject>
-                <imagedata fileref="img/javaPluginConfigurations.png"/>
+                <imagedata fileref="img/java-library-ignore-deprecated-main.png"/>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: We can't reuse the `java-library` graph 1:1 here, because it contains the `api` configuration, which the `java` plugin doesn't have.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 11]
```diff
@@ -375,10 +375,21 @@
                 </tr>
             </thead>
             <tr>
+                <td>implementation</td>
+                <td>api</td>
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: This is not the case. This relationship is only added by the `java-library` plugin and should thus be explained there.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 12]
```diff
@@ -361,8 +361,8 @@
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: I'm missing a general pointer towards the `java-library` plugin, which is what users will want to use when their project is a library (which is the 95% case).

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 10 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 9 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 11 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Done

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml (hunk 8 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: Done

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml (hunk 6 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: Done

File: subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml (hunk 7 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml]: Done

File: subprojects/docs/src/docs/userguide/javaPlugin.xml (hunk 12 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: Added to the end of the first paragraph.

File: subprojects/docs/src/docs/userguide/img/java-configurations.dot [hunk 13]
```diff
@@ -0,0 +1,21 @@
+digraph configurations {
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/img/java-configurations.dot]: We should show `apiElements` as well as the old `compile` and `runtime` configurations here, so people know where the soft-deprecated ones fit in.

File: subprojects/docs/src/docs/userguide/img/java-configurations.dot [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    rankdir = BT;
+   graph [ dpi = 100, fontname="Sans"];
+   node [fontname = "Sans"];
+   edge [fontname = "Sans"];
+
+   "compileClasspath" [shape="box", label=<<B>compileClasspath</B>(R)>, color="lightblue", bgcolor="lightblue", style="filled"]
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/img/java-configurations.dot]: The C/R labels aren't explained anywhere, I'd probably remove them.

File: subprojects/docs/src/docs/userguide/img/java-configurations.dot (hunk 13 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/img/java-configurations.dot]: Done

File: subprojects/docs/src/docs/userguide/img/java-configurations.dot (hunk 14 above)
Comment (eljobe) [inline subprojects/docs/src/docs/userguide/img/java-configurations.dot]: Done
